### PR TITLE
WFLY-10327 Upgrade Hibernate ORM to 5.3.0.Final and Hibernate Search to 5.10.0.CR1, squashed commits/changes from Sanne, Paul, Gail

### PR DESCRIPTION
**Pull request by scottmarlow**
https://issues.jboss.org/browse/WFLY-10327

### CODE REVIEW

File: feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml [hunk 1]
```diff
@@ -43,6 +51,7 @@
         <module name="org.hibernate.jipijapa-hibernate5" services="import"/>
         <module name="org.infinispan" services="import"/>
         <module name="org.infinispan.commons"/>
+        <module name="org.infinispan.hibernate-cache-commons" services="import"/>
```
Comment (jamezp) [inline feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml]: Why was this dependency added? It's also listed in the resources so it will be copied into this module.

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/jpa/hibernate/4/module.xml [hunk 2]
```diff
@@ -38,8 +38,6 @@
     </resources>
 
     <dependencies>
-        <module name="org.infinispan" services="import"/>
-        <module name="org.hibernate.infinispan" services="import"/>
```
Comment (jamezp) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/jpa/hibernate/4/module.xml]: Is there a reason this was removed? Will it work the same without these modules?

File: galleon-pack/pom.xml [hunk 3]
```diff
@@ -2248,8 +2270,8 @@
         </dependency>
 
         <dependency>
-            <groupId>org.hibernate</groupId>
-            <artifactId>hibernate-infinispan</artifactId>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-hibernate-cache-v53</artifactId>
```
Comment (jamezp) [inline galleon-pack/pom.xml]: Is the other version no longer used? These two dependencies need to be 100% compatible.

File: jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java [hunk 4]
```diff
@@ -82,35 +77,7 @@ public void addProviderProperties(Map properties, PersistenceUnitMetadata pu) {
...
-            HibernateSecondLevelCache.addSecondLevelCacheDependencies(pu.getProperties(), pu.getScopedPersistenceUnitName());
-            JPA_LOGGER.tracef("second level cache enabled for %s", pu.getScopedPersistenceUnitName());
-        } else {
-            JPA_LOGGER.tracef("second level cache disabled for %s, pu %s property = %s, pu.getSharedCacheMode = %s",
-                    pu.getScopedPersistenceUnitName(),
-                    SHARED_CACHE_MODE,
-                    sharedCacheMode,
-                    pu.getSharedCacheMode().toString());
-        }
+        JPA_LOGGER.tracef("second level cache not integrated for Hibernate 4.x - %s", pu.getScopedPersistenceUnitName());
```
Comment (jamezp) [inline jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java]: Why was this support removed? Having this as a trace message seems wrong as well.

File: jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/infinispan/InfinispanRegionFactory.java [hunk 5]
```diff
@@ -19,74 +19,19 @@
...
-    public static final String CACHE_CONTAINER = "hibernate.cache.infinispan.container";
-    public static final String DEFAULT_CACHE_CONTAINER = "hibernate";
-    public static final String CACHE_PRIVATE = "private";
-
-    private volatile Wrapper wrapper;
-
-    public InfinispanRegionFactory() {
-        super();
-    }
+public class InfinispanRegionFactory {
```
Comment (jamezp) [inline jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/infinispan/InfinispanRegionFactory.java]: What's the purpose of this change?

File: jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/infinispan/SharedInfinispanRegionFactory.java [hunk 6]
```diff
@@ -24,62 +24,16 @@
...
 /**
- * Infinispan-backed region factory that retrieves its cache manager from the Infinispan subsystem.
- * This is used for (JPA) container managed persistence contexts.
- * Each deployment application will use a unique Hibernate cache region name in the shared cache.
+ * Disabled 2lc.
  *
  * @author Paul Ferraro
  * @author Scott Marlow
  */
 public class SharedInfinispanRegionFactory extends InfinispanRegionFactory {
```
Comment (jamezp) [inline jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/infinispan/SharedInfinispanRegionFactory.java]: There seems to be some removal of functionality in general with Hibernate 4.3 support. I'll stop commenting on each thing I find, but we need an explanation on why this was removed :)

File: jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/InfinispanRegionFactory.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+package org.jboss.as.jpa.hibernate5.infinispan;
+
+/**
+ * Infinispan-backed region factory for use with standalone (i.e. non-JPA) Hibernate applications.
+ * @author Paul Ferraro
+ * @author Scott Marlow
+ * @deprecated Use {@link org.infinispan.hibernate.cache.v53.InfinispanRegionFactory} instead in conjunction with "hibernate.cache.infinispan.shared" set to false.
+ */
+@Deprecated
+public class InfinispanRegionFactory extends DeprecatedInfinispanRegionFactory {
```
Comment (jamezp) [inline jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/InfinispanRegionFactory.java]: Why are we adding a deprecated class?

File: jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/SharedInfinispanRegionFactory.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * Infinispan-backed region factory that retrieves its cache manager from the Infinispan subsystem.
+ * This is used for (JPA) container managed persistence contexts.
+ * Each deployment application will use a unique Hibernate cache region name in the shared cache.
+ *
+ * @author Paul Ferraro
+ * @author Scott Marlow
+ * @deprecated Use {@link org.infinispan.hibernate.cache.v53.InfinispanRegionFactory} instead.
+ */
+@Deprecated
+public class SharedInfinispanRegionFactory extends DeprecatedInfinispanRegionFactory {
```
Comment (jamezp) [inline jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/SharedInfinispanRegionFactory.java]: Same question about adding a deprecated implementation.

File: jpa/hibernate5-legacy/pom.xml [hunk 9]
```diff
@@ -43,6 +43,17 @@
             <artifactId>hibernate-core</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-hibernate-cache-spi</artifactId>
+            <scope>provided</scope>
```
Comment (jamezp) [inline jpa/hibernate5-legacy/pom.xml]: It his some kind of new bridge for caching?

File: jpa/hibernate5-legacy/pom.xml [hunk 10]
```diff
@@ -59,5 +70,11 @@
             <scope>provided</scope>
         </dependency>
 
+      <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-hibernate-cache-commons</artifactId>
```
Comment (jamezp) [inline jpa/hibernate5-legacy/pom.xml]: It his some kind of new bridge for caching?

File: jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java [hunk 11]
```diff
@@ -43,7 +45,7 @@
     @Override
     public void start(SessionFactoryOptions settings, Properties properties) throws CacheException {
         Logger.LOGGER.deprecatedRegionFactory(this.getClass().getName(), this.getClass().getSuperclass().getSuperclass().getName(), SHARED, this.shared);
-        properties.setProperty(SHARED, this.shared);
+        properties.put(SHARED, this.shared);
```
Comment (jamezp) [inline jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java]: What's the purpose of this change? These should do the same thing.

File: jpa/hibernate5_3/pom.xml [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+            <groupId>org.hibernate.common</groupId>
+            <artifactId>hibernate-commons-annotations</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-core</artifactId>
+            <scope>provided</scope>
+            <version>5.3.0.Final</version>
```
Comment (jamezp) [inline jpa/hibernate5_3/pom.xml]: We should be using a property not a hard-coded version.

File: jpa/hibernate5_3/pom.xml [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+    <name>jipijapa Hibernate 5.3.x (JPA 2.2) integration</name>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.hibernate</groupId>
+                <artifactId>hibernate-core</artifactId>
+                <scope>provided</scope>
+                <version>5.3.0.Final</version>
```
Comment (jamezp) [inline jpa/hibernate5_3/pom.xml]: We should be using a property not a hard-coded version.

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import org.jboss.logging.Logger;
+import org.jboss.logging.annotations.MessageLogger;
+
+/**
+ * JipiJapa message range is 20200-20299
+ * note: keep duplicate messages in sync between different sub-projects that use the same messages
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ * @author Scott Marlow
+ */
+@MessageLogger(projectCode = "JIPI")
```
Comment (jamezp) [inline jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java]: This should probably have it's own project code. It looks like the same project code is used in other loggers as well. This could lead to id conflicts.

**Review by jamezp — CHANGES_REQUESTED** [review]

**Comment by jamezp** [comment]
@scottmarlow @pferraro I'm concerned about the removal of the second level cache support for Hibernate 4.3. Mostly for compatibility reasons.

File: pom.xml [hunk 15]
```diff
@@ -278,15 +279,19 @@
         <!-- Hardcoded in module.xml for the dual EE7 and EE8 support -->
         <version.org.glassfish.javax.json-1.0>1.0.4</version.org.glassfish.javax.json-1.0>
         <version.org.glassfish.soteria>1.0</version.org.glassfish.soteria>
+        <!-- <version.org.hibernate53>5.3.0.Final</version.org.hibernate53> -->
+        <!-- Hardcoded in module.xml for the dual EE7 and EE8 support -->
```
[the remark this answers is not in the document]
Comment (Sanne) [reply pom.xml]: @kabir had already uncommented it. @scottmarlow did you lose his commit?

**Comment by Sanne** [comment]
@jamezp we don't support Hibernate 4.3 in EAP7 so I don't think that's a concern we need to act on. It has been long deprecated and people have been aware they should not use it, beyond the normal deprecation cycles.

File: pom.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (scottmarlow) [reply pom.xml]: uncommenting...

**Comment by jamezp** [comment]
Perfect that's what I wanted to know @Sanne

File: feature-pack/pom.xml [hunk 16]
```diff
@@ -2080,8 +2102,8 @@
         </dependency>
 
         <dependency>
-            <groupId>org.hibernate</groupId>
-            <artifactId>hibernate-infinispan</artifactId>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-hibernate-cache-v53</artifactId>
```
[the remark this answers is not in the document]
Comment (scottmarlow) [reply feature-pack/pom.xml]: Infinispan 9.2.x includes the equivalent of org.hibernate:hibernate-infinispan now, which we include in the org.infinispan.hibernate-cache module.  If we are running EE 7, we use org.infinispan:infinispan-hibernate-cache-v51, if we are running EE 8, we use org.infinispan:infinispan-hibernate-cache-v53.

File: feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml (hunk 1 above)
Comment (scottmarlow) [reply feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml]: hmm, yeah, weird for it to be a resource and self referenced.  should be no reason for that, must of missed it.

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/jpa/hibernate/4/module.xml (hunk 2 above)
Comment (scottmarlow) [reply feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/jpa/hibernate/4/module.xml]: Mostly because using the Infinispan 2lc with older versions of Hibernate shouldn't work due to WF using newer (major) versions of Infinispan.  We could revert this part of the change.

@Sanne any feedback on this?  Also see other comments about Search on this pr.

File: feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml (hunk 1 above)
Comment (scottmarlow) [reply feature-pack/src/main/resources/modules/system/layers/base/org/infinispan/hibernate-cache/main/module.xml]: removed.

File: galleon-pack/pom.xml (hunk 3 above)
Comment (scottmarlow) [reply galleon-pack/pom.xml]: Hmm, I wonder if galleon-pack/pom should instead use the infinispan-hibernate-cache-v51 artifact that ORM 5.1 uses.  I'll try switching galleon-pack/pom + featurepack/pom.xml

Also, as mentioned above, Infinispan 9.2.x includes the equivalent of org.hibernate:hibernate-infinispan now, which we include in the org.infinispan.hibernate-cache module. If we are running EE 7, we use org.infinispan:infinispan-hibernate-cache-v51, if we are running EE 8, we use org.infinispan:infinispan-hibernate-cache-v53.  infinispan-hibernate-cache-v51 might be a better artifact to reference since that is for EE 7 mode.

File: pom.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (Sanne) [reply pom.xml]: ah I see, that was before you switched the versions.

File: jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/InfinispanRegionFactory.java (hunk 7 above)
Comment (scottmarlow) [reply jpa/hibernate5-3-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/InfinispanRegionFactory.java]: for applications that reference this class by name, so they still deploy on WF13/WF14.

File: jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java (hunk 4 above)
Comment (jamezp) [reply jpa/hibernate4_1/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java]: We might want something more than a trace log message then. Probably a warning at least.

File: jpa/hibernate5-legacy/pom.xml (hunk 10 above)
Comment (scottmarlow) [reply jpa/hibernate5-legacy/pom.xml]: Same as above, for applications that reference the 2lc integration classes by name, so they still deploy on WF13/WF14. We don't need this module, but added it for app compat reasons.

File: jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java (hunk 11 above)
Comment (scottmarlow) [reply jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java]: Just part of code cleanup.

File: jpa/hibernate5_3/pom.xml (hunk 12 above)
Comment (scottmarlow) [reply jpa/hibernate5_3/pom.xml]: that didn't work with the component-matrix changes but maybe it will now work without component-matrix, trying...

Comment (scottmarlow) [reply jpa/hibernate5_3/pom.xml]: seems to build with ${version.org.hibernate53}, changing...

File: jpa/hibernate5_3/pom.xml (hunk 13 above)
Comment (scottmarlow) [reply jpa/hibernate5_3/pom.xml]: changing this one also to <version>${version.org.hibernate53}</version>

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java (hunk 14 above)
Comment (scottmarlow) [reply jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java]: Under wildfly/jpa, I think we mostly use JIPI for the jipijapa integration code.  I'm not against changing this project code...

File: feature-pack/pom.xml (hunk 16 above)
[the remark this answers is not in the document]
Comment (jamezp) [reply feature-pack/pom.xml]: Okay as long as it all works the same as it does now that should not be a problem.

File: jpa/hibernate4_3/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java [hunk 17]
```diff
@@ -103,34 +97,7 @@ public void addProviderProperties(Map properties, PersistenceUnitMetadata pu) {
...
-            HibernateSecondLevelCache.addSecondLevelCacheDependencies(pu.getProperties(), pu.getScopedPersistenceUnitName());
-            JPA_LOGGER.tracef("second level cache enabled for %s", pu.getScopedPersistenceUnitName());
-        } else {
-            JPA_LOGGER.tracef("second level cache disabled for %s, pu %s property = %s, pu.getSharedCacheMode = %s",
-                    pu.getScopedPersistenceUnitName(),
-                    SHARED_CACHE_MODE,
-                    sharedCacheMode,
-                    pu.getSharedCacheMode().toString());
-        }
+        JPA_LOGGER.tracef("second level cache not integrated for Hibernate 4.x - %s", pu.getScopedPersistenceUnitName());
```
Comment (jamezp) [inline jpa/hibernate4_3/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java]: This should probably be a warn message. Having it a trace doesn't inform the user unless they turn on trace logging which is unlikely.

File: jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java (hunk 11 above)
Comment (jamezp) [reply jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java]: I'm curious on how this is a cleanup though. It's using `put` on a `java.util.Properties` type. Using `setProperty()` seems more appropriate to me.

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java (hunk 14 above)
Comment (jamezp) [reply jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java]: The issue would be if you end up with duplicate id's in each logger that map to different messages.

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java [hunk 18]
```diff
@@ -38,14 +38,17 @@ protected RegionFactory resolveRegionFactory(Map configurationValues, ServiceReg
...
-        else {
+        } else if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
+             || "NONE".equals(jpa_shared_code_mode.toString())) {
             // explicitly disable 2lc cache
             return NoCachingRegionFactory.INSTANCE;
         }
+        else {
+            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
+            return super.resolveRegionFactory(configurationValues, registry);
+        }
```
Comment (jamezp) [inline jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java]: This looks weird. Why don't we just use something like:
```
if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
     || "NONE".equals(jpa_shared_code_mode.toString())) {
    // explicitly disable 2lc cache
    return NoCachingRegionFactory.INSTANCE;
}
configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
return super.resolveRegionFactory(configurationValues, registry);
```

Also what is the rule for "UNSPECIFIED"? Is that value allowed in Java EE 7?

**Review by jamezp — CHANGES_REQUESTED** [review]

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java (hunk 18 above)
Comment (jamezp) [reply jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java]: Actually I see in the comment now that 5.1 defaults to off. The if statement is overly complex IMO though.

File: jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java (hunk 11 above)
Comment (scottmarlow) [reply jpa/hibernate5-legacy/src/main/java/org/jboss/as/jpa/hibernate5/infinispan/DeprecatedInfinispanRegionFactory.java]: Sorry, I was referring to a different change that was already merged.  For this change, we can revert it, the equivalent class for ORM 5.3 integration is now passed a Map, so this code is kind of line of line identical with that but doesn't need to be.   I'll push this change in a few minutes after testing it.

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java (hunk 14 above)
Comment (scottmarlow) [reply jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/JpaLogger.java]: changing...

File: jpa/hibernate4_3/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java (hunk 17 above)
Comment (scottmarlow) [reply jpa/hibernate4_3/src/main/java/org/jboss/as/jpa/hibernate4/HibernatePersistenceProviderAdaptor.java]: changing to a warning.

File: jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java (hunk 18 above)
Comment (scottmarlow) [reply jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java]: If the Hibernate ORM use_second_level_cache configuration setting is specified by the application, that is our first choice.  

If the (JPA standard) jpa_shared_code_mode is specified as UNSPECIFIED, I current think we should disable the 2lc (which is what I did).  We might flip this later if possible, to default on.  If jpa_shared_code_mode is NONE, we will also disable 2lc.

Current code is:
```
        if(Boolean.parseBoolean((String)use_second_level_cache)) {
            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
            return super.resolveRegionFactory(configurationValues, registry);
        } else if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
             || "NONE".equals(jpa_shared_code_mode.toString())) {
            // explicitly disable 2lc cache
            return NoCachingRegionFactory.INSTANCE;
        }
        else {
            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
            return super.resolveRegionFactory(configurationValues, registry);
        }
```

Explanation of UNSPECIFIED from https://docs.oracle.com/javaee/7/api/javax/persistence/SharedCacheMode.html :
UNSPECIFIED - Caching behavior is undefined: provider-specific defaults may apply.
NONE - Caching is disabled for the persistence unit.

**Comment by pferraro** [comment]
@scottmarlow I've also linked https://issues.jboss.org/browse/WFLY-10393 to this PR.

**Comment by kabir** [comment]
Retest this please

**Review by jamezp — APPROVED** [review]
